src/engine/BoundsExplainer.h

```diff
@@ -0,0 +1,108 @@
+/*********************                                                        */
+/*! \file BoundsExplainer.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Omri Isac, Guy Katz
+ ** This file is part of the Marabou project.
+ ** Copyright (c) 2017-2022 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved. See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** [[ Add lengthier description here ]]
+ **/
+
+#ifndef __BoundsExplainer_h__
+#define __BoundsExplainer_h__
+#include "TableauRow.h"
+#include "vector"
+#include "SparseUnsortedList.h"
+
+/*
+  A class which encapsulates bounds explanations of all variables of a tableau
+*/
+class BoundsExplainer {
+public:
+	BoundsExplainer( unsigned varsNum, unsigned rowsNum );
+
+	~BoundsExplainer();
+
+	/*
+	 * Returns the number of rows
+	 */
+	unsigned getRowsNum() const;
+
+	/*
+ 	* Returns the number of variables
+ 	*/
+	unsigned getVarsNum() const;
+
+	/*
+	 * Returns a bound explanation
+	*/
+	const std::vector<double>& getExplanation( unsigned var, bool isUpper );
+
+	/*
+	 * Given a row, updates the values of the bound explanations of its lhs according to the row
+	*/
+	void updateBoundExplanation( const TableauRow& row, bool isUpper );
+
+	/*
+	 * Given a row, updates the values of the bound explanations of a var according to the row
+	*/
+	void updateBoundExplanation( const TableauRow& row, bool isUpper, unsigned varIndex );
+
+	/*
+	 * Given a row as SparseUnsortedList, updates the values of the bound explanations of a var according to the row
+	*/
+	void updateBoundExplanationSparse( const SparseUnsortedList& row, bool isUpper, unsigned var );
+
+	/*
+	 * Deep copies of other BoundsExplainer
+	 */
+	BoundsExplainer& operator=( const BoundsExplainer& other );
+
+	/*
+	 * Adds a zero explanation at the end, and a zero entry to all explanations
+	 */
+	void addVariable();
+
+	/*
+	 * Resets an explanation (to be a zero vec, represented by an empty vec)
+	 */
+	void resetExplanation( unsigned var, bool isUpper );
+
+	/*
+	 * Updates an explanation, without necessarily using the recursive rule
+	 */
+	void injectExplanation( const std::vector<double>& expl, unsigned var, bool isUpper );
+
+private:
+	unsigned _varsNum;
+	unsigned _rowsNum;
+	std::vector<std::vector<double>> _upperBoundsExplanations;
```

Comment:
Vector is our wrapper for std::vector
upperBoundsExplanations --> upperBoundExplanations